illumos-utils/src/opte/illumos/port_manager.rs

```diff
@@ -454,31 +351,239 @@ impl PortManager {
+            PortType::Guest { slot: nic.slot },
+            None,
+            vpc_cfg,
+            firewall_rules,
+            external_ip.is_some(),
+        )
+    }
+
+    pub fn create_svc_port(
+        &self,
+        svc_id: Uuid,
+        svc_name: &str,
+        mac: MacAddr6,
+        external_ip: IpAddr,
+        vni: u32,
+        firewall_rules: &[VpcFirewallRule],
+    ) -> Result<(Port, PortTicket), Error> {
+        // TODO: Assume each service zone has a single port.
+        let port_name = format!("{}_{}0", XDE_LINK_PREFIX, svc_name);
+
+        // TODO: Arbitrary subnet, should come from a "service" VPC.
+        let vpc_subnet: Ipv4Network = "10.0.0.0/24".parse().unwrap();
+        // Use first non-reserved IP, see RFD 21, section 2.2, table 1.
+        let private_ip = vpc_subnet.iter().nth(5).unwrap();
+        let gateway = Gateway::from_subnet(&vpc_subnet.into());
+
+        // Convert to OPTE types
+        let vpc_subnet = Ipv4Cidr::from(vpc_subnet);
+        let private_ip = private_ip.into();
+        let gateway_ip = match gateway.ip {
+            IpAddr::V4(ip) => ip.into(),
+            IpAddr::V6(_) => unreachable!("expected IPv4 gateway"),
+        };
+        let gateway_mac = MacAddr::from(gateway.mac.into_array());
+        let guest_mac = MacAddr::from(mac.into_array());
+        let external_ip = match external_ip {
+            IpAddr::V4(ip) => ip.into(),
+            IpAddr::V6(_) => {
+                return Err(opte_ioctl::Error::InvalidArgument(String::from(
+                    "IPv6 is not yet supported for external addresses",
+                ))
+                .into());
+            }
+        };
+
+        let snat = SNat4Cfg {
+            external_ip,
+            // TODO: Assumes IP is exlusive to this service.
+            ports: 0..=65535,
+        };
+
+        // TODO: Return error for invalid VNI.
+        let vni = Vni::new(vni).unwrap();
+
+        let vpc_cfg = VpcCfg {
+            ip_cfg: IpCfg::Ipv4(Ipv4Cfg {
+                vpc_subnet,
+                private_ip,
+                gateway_ip,
+                snat: Some(snat),
+                external_ips: Some(external_ip),
+            }),
+            guest_mac,
+            gateway_mac,
+            vni,
+            phys_ip: self.inner.underlay_ip.into(),
+            boundary_services: default_boundary_services(),
+            // TODO-remove: Part of the external IP hack.
+            //
+            // NOTE: This value of this flag is irrelevant, since the driver
+            // always overwrites it. The field itself is used in the `oxide-vpc`
+            // code though, to determine how to set up the ARP layer, which is
+            // why it's still here.
+            proxy_arp_enable: true,
+            phys_gw_mac: Some(MacAddr::from(
+                self.inner.gateway_mac.into_array(),
+            )),
+            // TODO-completeness (#2153): Plumb domain search list
+            domain_list: vec![],
```

Comment:
Is this relevant or desirable for service ports? I thought the intent of that issue was to provided guest DNS clients a domain list when doing resolution for _internal_ DNS names, e.g., of other instances. I don't know that we'd want the same thing for service ports, nor am I really sure what data we would provide here.

Reply:
As-is, probably not relevant? Just left the same comment for guest ports. Anyways, did some deduplication so there's no longer a service-specific TODO here anymore.